lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs

```diff
@@ -501,11 +572,94 @@ impl std::fmt::Display for FilesystemBackendError {
 
 impl std::error::Error for FilesystemBackendError {}
 
-type SharedDirectoryInfoResponseHandler = (
-    efs::DeviceCreateRequest,
-    fn(
-        &mut FilesystemBackend,
-        efs::DeviceCreateRequest,
-        tdp::SharedDirectoryInfoResponse,
-    ) -> PduResult<Option<RdpdrPdu>>,
+/// A macro for generating a response handler struct and function boilerplate for a given response type.
+///
+/// Call this like:
+///
+/// ```
+/// response_handler!(
+///   SharedDirectoryInfoResponseHandler,
+///   InfoResponseHandlerFunction,
+///   tdp::SharedDirectoryInfoResponse
+/// );
+/// ```
+///
+/// in order to generate the following:
+///
+/// ```
+/// type InfoResponseHandlerFunction = Box<
+///     dyn FnOnce(
+///             &mut FilesystemBackend,
+///             tdp::SharedDirectoryInfoResponse,
+///         ) -> PduResult<Option<RdpdrPdu>>
+///         + Send,
+/// >;
+///
+/// struct SharedDirectoryInfoResponseHandler {
+///     handler: InfoResponseHandlerFunction,
+/// }
+///
+/// impl SharedDirectoryInfoResponseHandler {
+///     fn new(handler: InfoResponseHandlerFunction) -> Self {
+///         Self { handler }
+///     }
+///
+///     fn call(
+///         self,
+///         this: &mut FilesystemBackend,
+///         res: tdp::SharedDirectoryInfoResponse,
+///     ) -> PduResult<Option<RdpdrPdu>> {
+///         (self.handler)(this, res)
+///     }
+/// }
+///
+/// impl std::fmt::Debug for SharedDirectoryInfoResponseHandler {
+///     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+///         write!(f, "<SharedDirectoryInfoResponseHandler>")
+///     }
+/// }
```

Comment:
The main purpose of this is here -- we need each item of `FilesystemBackend` to be `Debug`, hence we can't just place a `FnOnce` directly into `FilesystemBackend` (like we were doing in [its equivalent in the rdp-rs version of the code](https://github.com/gravitational/teleport/blob/68d8647cd61e0bedb376f131f74a9c4eb5d43da9/lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs#L81))

Reply:
I refactored it to this in https://github.com/gravitational/teleport/pull/34122

https://github.com/gravitational/teleport/pull/34122/commits/ad773013aa0f7c5ccb3b54fe3c91c7c260f72b6b

If there's a nicer non-macro way to do that I'm all ears, let's move this discussion to that PR.

Reply:
So we still use some macro there, right? 
Have you tried my proposal above? Is there any problem with it?

Reply:
Moved the discussion to https://github.com/gravitational/teleport/pull/34122#discussion_r1382522921